references/obj_detection/train_pytorch.py

```diff
@@ -0,0 +1,141 @@
+    for i in targets["boxes"]:
+        i[0] = round(i[0] * width)
+        i[2] = round(i[2] * width)
+        i[1] = round(i[1] * height)
+        i[2] = round(i[3] * height)
+    target_ = [{i: torch.tensor(j, dtype=torch.float32).to(device)} for i, j in targets.items()]
+    loss_dict = model_(x, target_)
+    loss = sum([l for k, l in loss_dict.items()])
+    return loss
+
+
+def evaluate(images, targets, val_metric, model_):
+    x, y = images.to(device), targets
+    targets = [torch.stack(j, dim=1)[0].to(device) for i, j in y.items()]
+    pq_metric, seg_quality, recall, precision = val_metric(targets=targets, model=model_, x=x)
+    return pq_metric, seg_quality, recall, precision
+
+
+def train_faster(optimizer, scheduler, train_dataloader, val_dataloader, mymodel: nn.Module, root_dir: str,
+                 label_path: str, num_epochs: int,
+                 val_metric):
+    """Training script
+    Args:
+        root_dir: location of image folder
+        label_path: location of label folder
+        num_epochs: number of epochs in training
+    Returns:
+        torch checkpoint
+    """
+    epch_pq = 0.0
+    epch_precision = 0.0
+    epch_recall = 0.0
+    epch_seg = 0.0
+    epch_train_loss = None
+    epch_loss = None
+    for i in range(num_epochs):
+        mean_loss = []
+        mean_pq = []
+        mean_precision = []
+        mean_recall = []
+        mean_seg = []
+        mymodel.train()
+        tk0 = tqdm(train_dataloader, total=int(len(train_dataloader)))
+        for images, targets in tk0:
+            tk0.set_description(f"Epoch:{i}_Train")
+            optimizer.zero_grad()
+            loss = fit_one_epoch(mymodel, images, targets, device)
+            mean_loss.append(loss)
+            loss.backward()
+            optimizer.step()
+            lo1 = loss.cpu().clone().detach().numpy()
+            tk0.set_postfix(loss=lo1, eph_loss=epch_train_loss, prec=epch_precision, rec=epch_recall, seg=epch_seg,
+                            pq=epch_pq)
+            time.sleep(0.1)
+        epch_loss = torch.sum(torch.stack(mean_loss, dim=0), dim=0) / len(train_dataloader)
+        epch_train_loss = epch_loss.cpu().clone().detach().numpy()
+        tk1 = tqdm(val_dataloader, total=int(len(val_dataloader)))
+        scheduler.step()
+        for images, targets in tk1:
+            tk1.set_description(f"Epoch:{i}_Val")
+            mymodel.eval()
+            pq_metric, seg_quality, recall, precision = evaluate(images, targets, val_metric, mymodel)
+            mean_pq.append(pq_metric)
+            mean_seg.append(seg_quality)
+            mean_recall.append(recall)
+            mean_precision.append(precision)
+        epch_pq = np.sum(np.array(mean_pq)) / len(val_dataloader)
+        epch_precision = np.sum(np.array(mean_precision)) / len(val_dataloader)
+        epch_recall = np.sum(np.array(mean_recall)) / len(val_dataloader)
+        epch_seg = np.sum(np.array(mean_seg)) / len(val_dataloader)
```

Comment:
this is supposed to be the evaluate function

Reply:
Thanks, changes has been made :)